processing/src/main/java/org/apache/druid/segment/join/HashJoinEngine.java

```diff
@@ -250,12 +244,32 @@ public boolean isDoneOrInterrupted()
         return isDone() || Thread.currentThread().isInterrupted();
       }
 
+      @Override
+      public void mark()
+      {
+        joinMarkId = joinColumnSelectorFactory.getRowId();
+      }
+
+      @Override
+      public void resetToMark()
+      {
+        leftCursor.reset();
+        joinMatcher.reset();
+        joinColumnSelectorFactory.resetRowId();
+        initialize();
+        while (!isDone() && joinColumnSelectorFactory.getRowId() < joinMarkId) {
```

Comment:
I wonder if this approach, where we fully reset and advance `leftCursor`, will work properly on top of mutable `leftCursor` such as `IncrementalIndexCursor`. Seems like it would be safer to pass through `mark` and `resetToMark` to the `leftCursor`.

Speaking of mutability, the rhs Joinable could also be mutable, like if it's a lookup backed by a Kafka topic. We don't snapshot those. That's also problematic for this logic, since it could make the `rowId` inconsistent post-reset.

I don't have a good solution in mind for problems like this. The simplest one I can think of right now is to side-step the problem by making it so not all cursors support `mark()` (for example, a join cursor might not support it) & then adjust TopN to use `runWithCardinalityUnknown` if the underlying cursor doesn't support `mark()` and `params.getNumValuesPerPass()` is less than `cardinality`.

Reply:
discussed offline but to update here, for now have decided to update these and loosen the contract so that mark/resetToMark only guarantee return to same timestamp since that is all topN needs, and also since stuff like lookup segments could change on iterating again, and added some tests about it